### PR TITLE
Supplementary checks for sufficient native token fees on MockRouter

### PR DESCRIPTION
**Pull request by jhweintraub**
## Motivation

Issue https://github.com/smartcontractkit/ccip/issues/755

## Solution

1. Added support in MockCCIPRouter contract for supplementary fee-checks in `ccipSend`. Function now includes toggle-able fees for users to check their contract interactions. If `feeToken` is the native gas token of the chain, supplementary check is included that `msg.value > getFee()`

2. constant `FEE_ENABLED` in MockRouter.sol can be enabled for testing with fees of `0.1 ether` for native gas Tokens and `1e18` otherwise.

### CODE REVIEW

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol [hunk 1]
```diff
@@ -1,5 +1,5 @@
 // SPDX-License-Identifier: MIT
-pragma solidity ^0.8.0;
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: Please undo this change, users might not be on 0.8.19 and would still want to use this contract

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol [hunk 2]
```diff
@@ -21,13 +21,17 @@ contract MockCCIPRouter is IRouter, IRouterClient {
   error InvalidAddress(bytes encodedAddress);
   error InvalidExtraArgsTag();
   error ReceiverError(bytes error);
+  error InsufficientNativeFeeTokens();
+  error InvalidNativeFeeTokens();
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: Let's use the exact errors used in the real router to make reverts realistic, they are in IRouterClient

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol [hunk 3]
```diff
@@ -117,8 +127,11 @@ contract MockCCIPRouter is IRouter, IRouterClient {
   }
 
   /// @notice Returns 0 as the fee is not supported in this mock contract.
-  function getFee(uint64, Client.EVM2AnyMessage memory) external pure returns (uint256 fee) {
-    return 0;
+  function getFee(uint64, Client.EVM2AnyMessage memory message) public pure returns (uint256 fee) {
+    if (FEES_ENABLED) {
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: Let's add a function `setFee` which sets a uint256 which is then used as the fee. That removed the need for the bool, and it's opt-in to use fees in the router. This function can then return that value

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol [hunk 4]
```diff
@@ -75,6 +79,12 @@ contract MockCCIPRouter is IRouter, IRouterClient {
     // We want to disallow sending to address(0) and to precompiles, which exist on address(1) through address(9).
     if (decodedReceiver > type(uint160).max || decodedReceiver < 10) revert InvalidAddress(message.receiver);
 
+    uint256 feeTokenAmount = getFee(0, message);
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: nit: even though it will be ignored, I'd prefer to pass in the dest chain selector over a 0-value

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol [hunk 5]
```diff
@@ -75,6 +79,12 @@ contract MockCCIPRouter is IRouter, IRouterClient {
...
+    if (feeTokenAmount != 0) {
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: nit: let's use the exact same if statements as the router does for easy readability

File: contracts/src/v0.8/ccip/test/mocks/test/MockRouterTest.t.sol [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    vm.expectRevert(MockCCIPRouter.InsufficientNativeFeeTokens.selector);
+    mockRouter.ccipSend(mockChainSelector, message);
+
+    //ccipSend with sufficient native tokens for fees
+    mockRouter.ccipSend{value: 0.1 ether}(mockChainSelector, message);
+
+    message.feeToken = address(1); //Set feeToken to something other than native asset
+    //Should revert because msg.value should be zero when feeToken is not native asset;
+    vm.expectRevert(MockCCIPRouter.InvalidNativeFeeTokens.selector);
+    mockRouter.ccipSend{value: 0.1 ether}(mockChainSelector, message);
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/mocks/test/MockRouterTest.t.sol:49]: One case is missing: the non-zero fee token with no value

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol (hunk 1 above)
Comment (matYang) [reply contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: to add some context, we use `^0.8.0` for contracts expected to be consumed by users

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol [hunk 7]
```diff
@@ -21,13 +21,17 @@ contract MockCCIPRouter is IRouter, IRouterClient {
...
+  error InsufficientNativeFeeTokens();
+  error InvalidNativeFeeTokens();
 
   event MessageExecuted(bytes32 messageId, uint64 sourceChainSelector, address offRamp, bytes32 calldataHash);
   event MsgExecuted(bool success, bytes retData, uint256 gasUsed);
 
   uint16 public constant GAS_FOR_CALL_EXACT_CHECK = 5_000;
   uint64 public constant DEFAULT_GAS_LIMIT = 200_000;
 
+  bool public constant FEES_ENABLED = true; //Enable to configure mock-fees for testing
```
Comment (matYang) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: ditto, this should be configurable

File: contracts/src/v0.8/ccip/test/mocks/test/MockRouterTest.t.sol [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+contract MockRouterTest is Test {
+  MockCCIPRouter public mockRouter;
+
+  function setUp() public {
+    mockRouter = new MockCCIPRouter();
+
+    deal(address(this), 100 ether);
+  }
+
+  function test_ccipSend_with_invalid_native_tokens_for_fee() public {
```
Comment (matYang) [inline contracts/src/v0.8/ccip/test/mocks/test/MockRouterTest.t.sol]: see other test examples, please stick with our test convention, tests are separated into `_Success` and `_Revert` cases.

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol [hunk 9]
```diff
@@ -67,14 +69,27 @@ contract MockCCIPRouter is IRouter, IRouterClient {
...
 
+    uint256 feeTokenAmount;
+    if (message.feeToken == address(0)) {
+      feeTokenAmount = getFee(destinationChainSelector, message);
+      if (msg.value < feeTokenAmount) revert InsufficientFeeTokenAmount();
+    } else {
+      if (msg.value > 0) revert InvalidMsgValue();
+    }
+
+    // if (feeTokenAmount != 0) {
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: Old code?

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol [hunk 10]
```diff
@@ -67,14 +69,27 @@ contract MockCCIPRouter is IRouter, IRouterClient {
...
-    uint64, // destinationChainSelector
+    uint64 destinationChainSelector,
     Client.EVM2AnyMessage calldata message
   ) external payable returns (bytes32) {
     if (message.receiver.length != 32) revert InvalidAddress(message.receiver);
     uint256 decodedReceiver = abi.decode(message.receiver, (uint256));
     // We want to disallow sending to address(0) and to precompiles, which exist on address(1) through address(9).
     if (decodedReceiver > type(uint160).max || decodedReceiver < 10) revert InvalidAddress(message.receiver);
 
+    uint256 feeTokenAmount;
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: doesn't have to be top level as we don't use it

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol [hunk 11]
```diff
@@ -117,8 +132,13 @@ contract MockCCIPRouter is IRouter, IRouterClient {
   }
 
   /// @notice Returns 0 as the fee is not supported in this mock contract.
-  function getFee(uint64, Client.EVM2AnyMessage memory) external pure returns (uint256 fee) {
-    return 0;
+  function getFee(uint64, Client.EVM2AnyMessage memory) public view returns (uint256) {
+    return MockFeeTokenAmount;
+  }
+
+  function setFee(uint256 _feeAmount) public returns (uint256) {
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: Shouldn't return anything & can be external + missing natspec

Comment (RensR) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: nit: we don't underscore args

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol [hunk 12]
```diff
@@ -28,6 +28,8 @@ contract MockCCIPRouter is IRouter, IRouterClient {
   uint16 public constant GAS_FOR_CALL_EXACT_CHECK = 5_000;
   uint64 public constant DEFAULT_GAS_LIMIT = 200_000;
 
+  uint256 public MockFeeTokenAmount; //use setFee() to change to non-zero to test fees
```
Comment (RensR) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: nit: style guide mentions all storage vars should have a `s_` prefix and be private/internal. Since this is already exposed through getFee we don't lose anything

File: contracts/src/v0.8/ccip/test/mocks/test/MockRouterTest.t.sol [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+    mockRouter.setFee(0.1 ether);
+
+    deal(address(this), 100 ether);
+
+    message.receiver = abi.encode(address(0x12345));
+    message.data = abi.encode("Hello World");
+  }
+
+  function test_ccipSendWithInsufficientNativeTokens_Revert() public {
+    //Should revert because did not include sufficient eth to pay for fees
```
Comment (matYang) [inline contracts/src/v0.8/ccip/test/mocks/test/MockRouterTest.t.sol:35]: nit: `//[space]comments`, same above and below

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol [hunk 14]
```diff
@@ -28,6 +28,8 @@ contract MockCCIPRouter is IRouter, IRouterClient {
   uint16 public constant GAS_FOR_CALL_EXACT_CHECK = 5_000;
   uint64 public constant DEFAULT_GAS_LIMIT = 200_000;
 
+  uint256 internal s_MockFeeTokenAmount; //use setFee() to change to non-zero to test fees
```
Comment (matYang) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol]: nit: `s_[lowercase]`

File: contracts/src/v0.8/ccip/test/mocks/MockRouter.sol [hunk 15]
```diff
@@ -67,14 +69,21 @@ contract MockCCIPRouter is IRouter, IRouterClient {
...
     if (message.receiver.length != 32) revert InvalidAddress(message.receiver);
     uint256 decodedReceiver = abi.decode(message.receiver, (uint256));
     // We want to disallow sending to address(0) and to precompiles, which exist on address(1) through address(9).
     if (decodedReceiver > type(uint160).max || decodedReceiver < 10) revert InvalidAddress(message.receiver);
 
+    if (message.feeToken == address(0)) {
+      uint256 feeTokenAmount = getFee(destinationChainSelector, message);
+      if (msg.value < feeTokenAmount) revert InsufficientFeeTokenAmount();
+    } else {
+      if (msg.value > 0) revert InvalidMsgValue();
```
Comment (matYang) [inline contracts/src/v0.8/ccip/test/mocks/MockRouter.sol:84]: I know it's not part of original scope, if user is sending non-native fee token, e.g. LINK, we'd want to make sure their test covers token approvals, can we add something like `IERC20(message.feeToken).safeTransferFrom(msg.sender, address(this), feeTokenAmount);` and corresponding tests

**Review by matYang — COMMENTED** [review]
quick note, ok for this one, going forward, this repo is public, don't put ticket numbers in description or branch names

**Review by RensR — APPROVED** [review]
